### PR TITLE
feature!: add maximum character limit to password length

### LINKED ISSUE #21433: Add a max length verification for the password and increase the limit to 100 characters
# Feature Issue

### User Story

As a user, I want the system to enforce a maximum password length of 100 characters during registration, login and password change flow **so that**  can crate secure passwords without exceeding the allowed limit and have consistent behavior across platforms.. 

### Description

Design for error message for desktop: https://www.figma.com/design/17fc13UBFvInrLgNUKJJg5/Kuba%E2%8E%9CDesktop?node-id=1023-298171&node-type=canvas&t=1Gmu9iQ8zkv57pux-0
https://www.figma.com/design/ibJOTPlNtIxESwS96vJb06/%F0%9F%91%A4-Profile-%7C-Desktop?node-id=7860-189803&node-type=frame&t=hnWzqq4UThCTihkJ-0

### Acceptance Criteria
**Max Length Validation:** The system should validate the password length during user registration and login, ensuring it does not exceed 100 characters.
**Error Message:** If the password exceeds 100 characters, an error message should be displayed, informing the user that the maximum length has been reached.
**UI Consistency:** The max length validation should be applied consistently across all platforms (desktop and mobile apps).
**Existing Passwords:** Existing users with passwords shorter than 100 characters should not be affected by this change.

### PR DESCRIPTION
**Pull request by seanstrom**
fixes #21433

### Summary

* This PR attempts to add some extra validation when creating or changing a password in the status mobile app. Now the password will not be allowed to exceed a length of 100 characters.

#### Platforms

- Android
- iOS

##### Functional

- Onboarding - Create Password
- Settings - Change Password

### Steps to test

- Open Status mobile app
- Create a new profile and proceed to the "Create profile password" screen
- Insert a password into the password input field that exceeds a 100 characters
- Check to see if password hint shows "Maximum 100 characters"
- After logging into the app, open the Settings screen
- Navigate to the "Password" section of settings
- Press on "Change password"
- Enter the old password and proceed to the "New Password" screen
- Insert a password into the password input field that exceeds a 100 characters
- Check to see if password hint shows "Maximum 100 characters" 

##### Create Profile Password

https://github.com/user-attachments/assets/340fb119-ee2a-41ba-9096-4f4117fb1e78

##### Change Profile Password

https://github.com/user-attachments/assets/0a9dff3e-331a-49ba-bd89-b61adf9ac2ed

status: ready

### CODE REVIEW

File: src/status_im/common/password_with_hint/view.cljs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+               :blur? true))]
+   [rn/view {:style style/info-message}
+    (when shown?
+      [quo/info-message
+       (cond-> {:status status
+                :size   :default}
+         (not= :success status) (assoc :icon :i/info)
+         (= :success status)    (assoc :icon :i/check-circle)
+         (= :default status)    (assoc :color colors/white-70-blur))
+       text])]])
```
Comment (seanstrom) [inline src/status_im/common/password_with_hint/view.cljs]: Here we're creating a common component for both the Create Password and Change Password screens.

File: src/status_im/contexts/onboarding/create_password/view.cljs [hunk 2]
```diff
@@ -58,16 +41,21 @@
...
-                        :shown  true}
+     [password-with-hint/view
+      {:hint           (if (not password-short-enough?)
+                         {:text   (i18n/label
+                                   :t/password-creation-max-length-hint)
+                          :status :error
+                          :shown? true}
+                         {:text   (i18n/label :t/password-creation-hint)
+                          :status hint-1-status
+                          :shown? true})
```
Comment (seanstrom) [inline src/status_im/contexts/onboarding/create_password/view.cljs:52]: Here we're showing the "maximum 100 characters" error for the Create Password form.

File: src/status_im/contexts/profile/settings/screens/password/change_password/new_password_form.cljs [hunk 3]
```diff
@@ -115,17 +101,21 @@
...
-                         :status (if long-enough? :success :default)
-                         :shown? true}
+      [password-with-hint/view
+       {:hint           (if (not short-enough?)
+                          {:text   (i18n/label :t/password-creation-max-length-hint)
+                           :status :error
+                           :shown? true}
+                          {:text   (i18n/label :t/password-creation-hint)
+                           :status (if long-enough? :success :default)
+                           :shown? true})
```
Comment (seanstrom) [inline src/status_im/contexts/profile/settings/screens/password/change_password/new_password_form.cljs:93]: Here we're showing the "maximum 100 characters" error for the Change Password form.

**Review by seanstrom — COMMENTED** [review]
Self-Review 📑

File: src/status_im/common/password_with_hint/view.cljs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
```
Comment (seanstrom) [inline src/status_im/common/password_with_hint/view.cljs]: Here we're creating a common component for both the Create Password and Change Password screens.

**Review by Parveshdhull — APPROVED** [review]

**Comment by VolodLytvynenko** [comment]
hi @seanstrom, thank you for PR. take a look found issues:

## ISSUE 1: "Very weak" validation is shown on the 'create profile password' page during onboarding in case when the password field is empty

### Steps:
Go to onboarding password creation
### Actual result:
- The "Very weak" validation label is shown even when the password field is empty.
- Once at least one character is entered, "Very weak" disappears.
image

### Expected result:
- "Very weak" validation label should not be shown if the password field is empty.
- The "Very weak" label should appear when 1-2 characters are entered.

https://www.figma.com/design/o4qG1bnFyuyFOvHQVGgeFY/Onboarding-for-Mobile?node-id=12860-104946&node-type=frame&t=9YKldNHYn84kodx1-0

**Comment by VolodLytvynenko** [comment]
I've just rethought the [issue](https://github.com/status-im/status-mobile/pull/21593#issuecomment-2462112662) and [asked](https://discord.com/channels/1210237582470807632/1217173176086302821/1304070149833363486) the design team to match expected result

**Comment by VolodLytvynenko** [comment]
## PR_ISSUE 2:  "Passwords do not match" validation is not shown on the 'create profile password' page during onboarding 

### Steps:
1. Enter different passwords within create and repeat password fields

### Actual result:
image

### Expected result:
"Passwords do not match" validation is shown

image

https://www.figma.com/design/o4qG1bnFyuyFOvHQVGgeFY/Onboarding-for-Mobile?node-id=12860-105248&node-type=frame&t=9YKldNHYn84kodx1-0


### Devices:
- Pixel 7a, Android 13

File: src/status_im/common/password_with_hint/view.cljs [hunk 5]
```diff
@@ -0,0 +1,24 @@
+(ns status-im.common.password-with-hint.view
+  (:require
+    [quo.core :as quo]
+    [quo.foundations.colors :as colors]
+    [react-native.core :as rn]
+    [status-im.common.password-with-hint.style :as style]))
+
+(defn view
+  [{{:keys [text status shown?]} :hint :as input-props}]
+  [rn/view
```
Comment (ilmotta) [inline src/status_im/common/password_with_hint/view.cljs]: Could it be a fragment instead of `view`?

**Comment by VolodLytvynenko** [comment]
## PR_ISSUE 3: Password field allows input exceeding 100 characters on 'Create Profile Password' Page and proceeds with such amount

### Steps:
1. Enter in create password and repeat password the value which exceeding 100 characters
2. Go to the next page

### Actual result:
Users can proceed with a password exceeding 100 characters.
image

### Expected result:
Confirm password buttons is disabled when max 100 chars validation is enabled

### Devices:
- Pixel 7a, Android 13

**Review by ilmotta — APPROVED** [review]

File: src/status_im/common/password_with_hint/view.cljs (hunk 5 above)
Comment (seanstrom) [reply src/status_im/common/password_with_hint/view.cljs]: Yeah good idea! I'll change it to a fragment 👍

**Comment by seanstrom** [comment]
@VolodLytvynenko thanks for finding these issues 🙌 

For [issue 2](https://github.com/status-im/status-mobile/pull/21593#issuecomment-2462252826) and [issue 3 ](https://github.com/status-im/status-mobile/pull/21593#issuecomment-2462322672), I've pushed up some changes to fix those issues.

Let me know if [issue 1](https://github.com/status-im/status-mobile/pull/21593#issuecomment-2462112662) should be addressed in this PR please 🙏

**Comment by VolodLytvynenko** [comment]
Hi @seanstrom, thank you for fixes. And sorry for the PR's back-and-forth. I think it's best to fix Issue 1 within this PR. We don’t know when the validation simplification will be done, and merging as-is would mean showing 'Very Weak' by default on the create profile password page, which seems worse than the current issue on the latest develop where 'Very Strong' is not shown at all

Also, the current PR has different validation logic for 'create profile password' and 'change password':

For now:
On `Create profile password` page:

`Very Weak - shown by default`
`Weak - 1 condition met`
`OK - 2 conditions met`
`Strong - 3 conditions met`
`Very Strong - 4 conditions met`

While on `Change password`:

`Very Weak - 1 condition met`
`Weak - 2 conditions met`
`OK - 3 conditions met`
`Strong - 4 conditions met`
`Very Strong - never shown`

Neither setup is ideal, but aligning them with a consistent logic would be better. 

**Expected result:**
The logic of validation of a strong password follows the same rules as on the latest nightly
`Very Weak - 1 condition met`
`Weak - 2 conditions met`
`OK - 3 conditions met`
`Strong - 4 conditions met`
`Very Strong - never shown`

**Comment by ilmotta** [comment]
Hey @VolodLytvynenko @seanstrom, I think this password strength validator leaves a lot to be desired. For example, a password `aA1!` is shown as strong in Status. Check out https://www.passwordmonster.com/, over there we see this password can be cracked instantly and it is shown as very weak. Our implementation is leading the user in the wrong direction, which is bad for security.

I would remove this feature completely from the onboarding and only keep the basic length validation, but to align this point would take approvals from other folks. If you would like to see this discussion move forward please give a thumbs up and I'll gladly try to help.

Edit: this separate discussion shouldn't block this PR 🚀

**Comment by seanstrom** [comment]
@VolodLytvynenko Sounds good, about making them consistent and not showing "Very weak" by default 👍

I've pushed a change that should make them consistent for both change-password and create-password screens. Now both of them will not show "Very weak" by default. Let me know what you think 🙏

**Comment by seanstrom** [comment]
@ilmotta we may be able to create a more nuanced calculation for password strength. Currently we only count how many validations have passed, but we could try to create something that only follows some of the guidelines mentioned [here](https://discord.com/channels/1210237582470807632/1217173176086302821/1304106241639059557).

Overall I'm okay with removing a misleading feature, and if we can prioritise a better feature that would be nice too 🙌

**Comment by ilmotta** [comment]
The following is an interesting discussion from Bitwarden. Even more sophisticated calculations can be very misleading.

**Do you use password strength indicators? Complete this survey** https://community.bitwarden.com/t/do-you-use-password-strength-indicators-complete-this-survey/62738/2

Next week I'll try to take the discussion somewhere else and hear other people's opinions. Similar to how we discussed about identity rings and eventually descoped it completely from the product. Now, better than ever is a good chance to understand why things were built the way they were and improve.

**Comment by ilmotta** [comment]
@seanstrom @VolodLytvynenko In the desktop app, the strength validator is somewhat useful, but in mobile it's quite misleading/wrong. For example, in desktop, the password `aA1!` is considered _very weak_, in mobile it's considered _strong_.

I'll create a new issue about this feature parity problem 👍🏼 It's going to be a low priority feature to work on for 2.32, even though it's related to the onboarding.

**Comment by VolodLytvynenko** [comment]
hi @seanstrom Thank you for the fixes. Unfortunately, additional issue is found

## PR_ISSUE 4: "Maximum 100 chars" validation is shown by default on create "password password" and "change password" pages

### Steps:
1. Open "create profile password" page
2. Open "change profile password" page

### Actual result:
"Maximum 100 chars" validation is shown by default
image

### Expected result:
The "Maximum 100 chars" validation message should only be displayed after the user enters more than 100 characters.

### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

**Comment by seanstrom** [comment]
Hey @VolodLytvynenko thank you for catching that issue, I've pushed another fix that should show the correct password hint before typing in the password.

**Comment by VolodLytvynenko** [comment]
hi @seanstrom thank you for PR. No issues from my side. PR is ready to be merged

**Pull request merged**